File: code/backend/php/src/Core/Port/Story/StoryReadPort.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Port\Story;

interface StoryReadPort
{
    public function findPublicStories(): array;

    public function findAllStories(): array;

    public function findStoryByUuid(string $uuid): ?array;

    public function findTextsForStory(int $storyId): array;

    public function findDifficultiesForStory(int $storyId): array;

    public function countLocationsForStory(int $storyId): int;

    public function countEventsForStory(int $storyId): int;

    public function countItemsForStory(int $storyId): int;

    public function findUniqueCategories(): array;

    public function findUniqueGroups(): array;

    public function findStoriesByCategory(string $category): array;

    public function findStoriesByGroup(string $group): array;

    public function findClassesForStory(int $storyId): array;

    public function findCharacterTemplatesForStory(int $storyId): array;

    public function findTraitsForStory(int $storyId): array;

    public function findCardForStory(int $storyId, int $cardId): ?array;

    public function findCardByStoryIdAndUuid(int $storyId, string $uuid): ?array;

    public function findTextByStoryIdTextAndLang(int $storyId, int $idText, string $lang): ?array;

    public function findCreatorByStoryIdAndUuid(int $storyId, string $uuid): ?array;

    public function findCreatorsForStory(int $storyId): array;

    // Step 17: Generic CRUD read support

    public function findLocationsForStory(int $storyId): array;

    public function findEventsForStory(int $storyId): array;

    public function findItemsForStory(int $storyId): array;

    public function findCardsForStory(int $storyId): array;

    public function findEntityByStoryAndUuid(int $storyId, string $tableName, string $uuid): ?array;

    public function findEntitiesByStory(int $storyId, string $tableName): array;
}

File: code/backend/python/app/core/ports/story/story_read_port.py
from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional

class StoryReadPort(ABC):
    @abstractmethod
    def find_public_stories(self) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_all_stories(self) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_story_by_uuid(self, uuid: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_texts_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_difficulties_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def count_locations_for_story(self, story_id: int) -> int:
        pass

    @abstractmethod
    def count_events_for_story(self, story_id: int) -> int:
        pass

    @abstractmethod
    def count_items_for_story(self, story_id: int) -> int:
        pass

    @abstractmethod
    def find_unique_categories(self) -> List[str]:
        pass

    @abstractmethod
    def find_unique_groups(self) -> List[str]:
        pass

    @abstractmethod
    def find_stories_by_category(self, category: str) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_stories_by_group(self, group: str) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_classes_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_character_templates_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_traits_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_card_for_story(self, story_id: int, card_id: int) -> Optional[Dict[str, Any]]:
        pass

    # Step 16: Content detail queries

    @abstractmethod
    def find_card_by_story_id_and_uuid(self, story_id: int, uuid: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_text_by_story_id_text_and_lang(self, story_id: int, id_text: int, lang: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_creator_by_story_id_and_uuid(self, story_id: int, uuid: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def find_creators_for_story(self, story_id: int) -> List[Dict[str, Any]]:
        pass
